test-js/unit/pentaho/util/promise.Spec.js

```diff
@@ -0,0 +1,316 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+define([
+  "pentaho/util/promise",
+  "pentaho/util/error"
+], function(promiseUtil, error) {
+
+  "use strict";
+
+  /*global describe:false, it:false, expect:false, beforeEach:false, afterEach:false, spyOn:false, jasmine:false*/
+
+  describe("pentaho.util.promise -", function() {
+
+    describe("wrapCall(fun(, ctx))", function() {
+
+      it("should throw if fun is not specified", function() {
+        function expectIt(args) {
+          expect(function() {
+
+            promiseUtil.wrapCall.apply(promiseUtil, args);
+
+          }).toThrowError(error.argRequired("fun").message);
+        }
+
+        expectIt([]);
+        expectIt([null]);
+        expectIt([undefined]);
+        expectIt([null, {}]);
+        expectIt([undefined, {}]);
+      });
+
+      it("should return a promise", function() {
+        var result = promiseUtil.wrapCall(function() {});
+        expect(result instanceof Promise).toBe(true);
+      });
+
+      it("should call fun synchronously", function() {
+        var called = false;
+        promiseUtil.wrapCall(function() { called = true; });
+        expect(called).toBe(true);
+      });
+
+      it("should resolve with the function's return value", function(done) {
+        var result = {};
+        promiseUtil.wrapCall(function() { return result; })
+            .then(function(value) {
+              expect(value).toBe(result);
+              done();
+            }, done.fail);
+      });
+
+      it("should reject with the error thrown by the function (and not throw sync)", function(done) {
+        var error = {};
+        promiseUtil.wrapCall(function() { throw error; })
+            .then(function() {
+              done.fail();
+            }, function(reason) {
+              expect(reason).toBe(error);
+              done();
+            });
+      });
+
+      it("should call fun on the given ctx object", function() {
+        var fun = jasmine.createSpy("promiseWrapCall");
+        var ctx = {};
+        promiseUtil.wrapCall(fun, ctx);
+        expect(fun.calls.count()).toBe(1);
+        expect(fun.calls.first().object).toBe(ctx);
+      });
+    });
+
+    describe("require(deps, localRequire)", function() {
+      it("should throw when `deps` is not specified", function() {
+        function expectIt(args) {
+          expect(function() {
+
+            promiseUtil.require.apply(promiseUtil, args);
+
+          }).toThrowError(error.argRequired("deps").message);
+        }
+
+        expectIt([]);
+        expectIt([null]);
+        expectIt([undefined]);
+      });
+
+      it("should fulfill to a single value when `deps` is a string", function(done) {
+        var mid = "pentaho/tests/promise/require/test1";
+        require.undef(mid);
```

Comment:
That's an issue I haven't been able to solve well yet. I've tried using Squire a number of times — actually every time I had to test async stuff with AMD modules in pentaho/type. The problem is that it goes completely crazy and totally unrelated tests start breaking everywhere. I was able to use Squire in geo, I remember, but in common-ui, I've been unsuccessful. So I turned to doing it directly.
For simple cases, that don't have dependencies, this method has no hidden dangers.
The alternative would be to create requirejs contexts ourselves, but we would be leaking memory that way, something which come become a problem, and doesn't seem right any way.
The best solution would be one in which the requirejs library itself would provide a testing library.